proposals/csharp-10.0/parameterless-struct-constructors.md

```diff
@@ -0,0 +1,106 @@
+# Parameterless struct constructors
+
+## Summary
+
+Support explicit parameterless constructors for value types.
+
+## Proposal
+
+If the parameterless constructor is less accessible than the value type, a warning is reported and the constructor is ignored when inaccessible.
+```csharp
+struct NoConstructor
+{
+    // ok
+}
+struct PublicConstructor
+{
+    internal PublicConstructor() { } // ok: always accessible
+}
+struct PrivateConstructor
+{
+    private PrivateConstructor() { } // warning: parameterless .ctor will be ignored if inaccessible
+}
+```
+
+`default(S)` ignores any parameterless constructor and generates a zeroed `S`.
+```csharp
+NoConstructor x = default;      // ok
+PublicConstructor y = default;  // ok
+PrivateConstructor z = default; // ok
+```
+
+For a type parameter `T` with a `new()` constraint, `new T()` is emitted as a call to `System.Activator.CreateInstance<T>()` which ignores any explicit parameterless constructor.
+A warning is reported if there is an accessible parameterless constructor.
+
+_Should `Activator.CreateInstance()` call the parameterless constructor if available and accessible?_
+```csharp
+static T Create<T>() where T : new() => new T();
+
+x = Create<NoConstructor>();      // ok
+y = Create<PublicConstructor>();  // warning: constructor ignored
+z = Create<PrivateConstructor>(); // ok
```

Comment:
This should warn.

Reply:
Scratch that, this should error, like we do for reference types today. It will throw at runtime. As an example:

```il
.assembly 'test' {}

.class private auto ansi abstract sealed beforefieldinit '<Program>$'
    extends [mscorlib]System.Object
{
    .custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
        01 00 00 00
    )
    // Methods
    .method private hidebysig static 
        void '<Main>$' (
            string[] args
        ) cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 17 (0x11)
        .maxstack 8
        .entrypoint

        IL_0000: ldtoken S
        IL_0005: call class [mscorlib]System.Type [mscorlib]System.Type::GetTypeFromHandle(valuetype [mscorlib]System.RuntimeTypeHandle)
        IL_000a: call object [mscorlib]System.Activator::CreateInstance(class [mscorlib]System.Type)
        IL_000f: pop
        IL_0010: ret
    } // end of method '<Program>$'::'<Main>$'

} // end of class <Program>$

.class private sequential ansi sealed beforefieldinit S
    extends [mscorlib]System.ValueType
{
    .pack 0
    .size 1
    // Methods
    .method assembly hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x2062
        // Code size 17 (0x11)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [mscorlib]System.ValueType::.ctor()
        IL_0006: ldstr "Hello world!"
        IL_000b: call void [mscorlib]System.Console::WriteLine(string)
        IL_0010: ret
    } // end of method S::.ctor
} // end of class S
```

Prints:

```
Unhandled Exception: System.MissingMethodException: No parameterless constructor defined for this object.
   at System.RuntimeTypeHandle.CreateInstance(RuntimeType type, Boolean publicOnly, Boolean noCheck, Boolean& canBeCached, RuntimeMethodHandleInternal& ctor, Boolean& bNeedSecurityCheck)
   at System.RuntimeType.CreateInstanceSlow(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, Boolean nonPublic)
   at System.Activator.CreateInstance(Type type)
   at <Program>$.<Main>$(String[] args)
```

Reply:
(The sample works if you change `assembly` to `public` on the ctor, if that wasn't clear).